oximeter/db/src/client.rs

```diff
@@ -1733,611 +1658,932 @@ mod tests {
+            measurement_table,
+            crate::DATABASE_SELECT_FORMAT,
         );
+        let body = client
+            .execute_with_body(select_sql)
+            .await
+            .expect("Failed to select measurement row");
+        let actual_row: serde_json::Value = serde_json::from_str(&body)
+            .expect("Failed to parse measurement row JSON");
+        println!("{actual_row:?}");
+        println!("{inserted_row:?}");
         assert_eq!(
-            result.items[0].metric, last_timeseries.metric,
-            "Paginating should pick up where it left off"
+            actual_row, inserted_row,
+            "Actual and expected measurement rows do not match"
         );
-        db.cleanup().await.expect("Failed to cleanup database");
+        client.wipe_single_node_db().await?;
+        logctx.cleanup_successful();
+        Ok(())
+    }
+
+    #[tokio::test]
+    async fn test_build_replicated() {
+        let log = slog::Logger::root(slog::Discard, o!());
+
+        let mut cluster = ClickHouseCluster::new()
+            .await
+            .expect("Failed to initialise ClickHouse Cluster");
+
+        // Create database in node 1
+        let client_1 = Client::new(cluster.replica_1.address.unwrap(), &log);
+        assert!(client_1.is_oximeter_cluster().await.unwrap());
+        client_1
+            .init_replicated_db()
+            .await
+            .expect("Failed to initialize timeseries database");
+
+        // Wait to make sure data has been synchronised.
+        // TODO(https://github.com/oxidecomputer/omicron/issues/4001): Waiting for 5 secs is a bit sloppy,
+        // come up with a better way to do this.
+        sleep(Duration::from_secs(5)).await;
+
+        // Verify database exists in node 2
+        let client_2 = Client::new(cluster.replica_2.address.unwrap(), &log);
+        assert!(client_2.is_oximeter_cluster().await.unwrap());
+        let sql = String::from("SHOW DATABASES FORMAT JSONEachRow;");
+        let result = client_2.execute_with_body(sql).await.unwrap();
+        assert!(result.contains("oximeter"));
+
+        // Insert row into one of the tables
+        let sql = String::from(
+            "INSERT INTO oximeter.measurements_string (datum) VALUES ('hiya');",
+        );
+        client_2.execute_with_body(sql).await.unwrap();
+
+        let sql = String::from(
+            "SELECT * FROM oximeter.measurements_string FORMAT JSONEachRow;",
+        );
+        let result = client_2.execute_with_body(sql.clone()).await.unwrap();
+        assert!(result.contains("hiya"));
+
+        // TODO(https://github.com/oxidecomputer/omicron/issues/4001): With distributed
+        // engine, it can take a long time to sync the data. This means it's tricky to
+        // test that the data exists on both nodes.
+
+        cluster
+            .keeper_1
+            .cleanup()
+            .await
+            .expect("Failed to cleanup ClickHouse keeper 1");
```

Comment:
Another place an array would help clean things up, since we could iterate over it here.